leakcanary-analyzer/src/main/java/com/squareup/leakcanary/HeapAnalyzer.java

```diff
@@ -90,6 +95,28 @@ public AnalysisResult checkForLeak(File heapDumpFile, String referenceKey) {
     }
   }
 
+  void dedupGcRoots(Snapshot snapshot) {
+    // THashMap has smaller footprint than HashMap
+    final THashMap<String, RootObj> dedupMap = new THashMap<>();
+
+    final List<RootObj> gcRoots = (ArrayList) snapshot.getGCRoots();
+    for (RootObj root : gcRoots) {
+      String id = root.toString();
```

Comment:
`id` => `key`. Also, nothing here conveys that we're building a unique key based `id` and `type`. Should we do it manually here instead of relying on toString()? Would be more clear. Could also be more efficient (not using String.format, and maybe even using masking to combine a type int and the id)